deepmd/utils/data_system.py

```diff
@@ -508,6 +514,626 @@ def _prob_sys_size_ext(self, keywords):
+                     ) )
+        log.info("--------------------------------------------------------------------------------------")
+
+    def _make_auto_bs(self, rule) :
+        bs = []
+        for ii in self.data_systems:
+            ni = ii.get_natoms()
+            bsi = rule // ni
+            if bsi * ni < rule:
+                bsi += 1
+            bs.append(bsi)
+        return bs
+
+    # ! added by Marián Rynik
+    def _make_auto_ts(self, percent):
+        ts = []
+        for ii in range(self.nsystems):
+            ni = self.batch_size[ii] * self.nbatches[ii]
+            tsi = int(ni * percent / 100)
+            ts.append(tsi)
+
+        return ts
+
+    def _check_type_map_consistency(self, type_map_list):
+        ret = []
+        for ii in type_map_list:
+            if ii is not None:
+                min_len = min([len(ii), len(ret)])
+                for idx in range(min_len) :
+                    if ii[idx] != ret[idx] :
+                        raise RuntimeError('inconsistent type map: %s %s' % (str(ret), str(ii)))
+                if len(ii) > len(ret) :
+                    ret = ii
+        return ret
+
+    def _process_sys_probs(self, sys_probs) :
+        sys_probs = np.array(sys_probs)
+        type_filter = sys_probs >= 0
+        assigned_sum_prob = np.sum(type_filter * sys_probs)
+        assert assigned_sum_prob <= 1, "the sum of assigned probability should be less than 1"
+        rest_sum_prob = 1. - assigned_sum_prob
+        if rest_sum_prob != 0 :
+            rest_nbatch = (1 - type_filter) * self.nbatches
+            rest_prob = rest_sum_prob * rest_nbatch / np.sum(rest_nbatch)
+            ret_prob = rest_prob + type_filter * sys_probs
+        else :
+            ret_prob = sys_probs
+        assert np.sum(ret_prob) == 1, "sum of probs should be 1"
+        return ret_prob
+    
+    def _prob_sys_size_ext(self, keywords):
+        block_str = keywords.split(';')[1:]
+        block_stt = []
+        block_end = []
+        block_weights = []
+        for ii in block_str:
+            stt = int(ii.split(':')[0])
+            end = int(ii.split(':')[1])
+            weight = float(ii.split(':')[2])
+            assert(weight >= 0), "the weight of a block should be no less than 0"
+            block_stt.append(stt)
+            block_end.append(end)
+            block_weights.append(weight)
+        nblocks = len(block_str)
+        block_probs = np.array(block_weights) / np.sum(block_weights)
+        sys_probs = np.zeros([self.get_nsystems()])
+        for ii in range(nblocks):
+            nbatch_block = self.nbatches[block_stt[ii]:block_end[ii]]
+            tmp_prob = [float(i) for i in nbatch_block] / np.sum(nbatch_block)
+            sys_probs[block_stt[ii]:block_end[ii]] = tmp_prob * block_probs[ii]
+        return sys_probs
+
+class DeepmdDataDocker() :
+    """
+    Class for manipulating many data dockers. 
+    It is implemented with the help of DeepmdData
+    """
```

Comment:
If most methods are copied from `DeepmdData`, why not inherit `DeepmdData`?

Reply:
Either inherit from `DeepmdData` or implement with `DeepmdData` object as a private member.